cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,409 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#include <cub/config.cuh>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+_CCCL_SUPPRESS_DEPRECATED_PUSH
+#include <cuda/std/functional>
+_CCCL_SUPPRESS_DEPRECATED_POP
+
+#include <cub/device/dispatch/dispatch_reduce.cuh>
+#include <cub/iterator/arg_index_input_iterator.cuh>
+#include <cub/iterator/constant_input_iterator.cuh>
+
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/tabulate_output_iterator.h>
+
+#include <cuda/std/type_traits>
+
+#ifndef _CCCL_DOXYGEN_INVOKED // Do not document
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+namespace reduce
+{
+
+template <typename GlobalAccumT, typename PromoteToGlobalOpT, typename GlobalReductionOpT, typename FinalResultOutIteratorT>
+struct accumulating_transform_output_op
+{
+  bool first_partition;
+  bool last_partition;
+
+  // We use a double-buffer to make assignment idempotent (i.e., allow potential repeated assignment)
+  GlobalAccumT* d_previous_aggregate;
+  GlobalAccumT* d_aggregate_out;
+
+  // Output iterator to which the final result of type `GlobalAccumT` across all partitions will be assigned
+  FinalResultOutIteratorT d_out;
+
+  // Unary promotion operator type that is used to transform a per-partition result to a global result
+  PromoteToGlobalOpT promote_op;
+
+  // Reduction operation
+  GlobalReductionOpT reduce_op;
+
+  template <typename IndexT, typename AccumT>
+  _CCCL_DEVICE _CCCL_FORCEINLINE void operator()(IndexT, AccumT per_partition_aggregate)
+  {
+    // Add this partitions aggregate to the global aggregate
+    if (first_partition)
+    {
+      *d_aggregate_out = promote_op(per_partition_aggregate);
+    }
+    else
+    {
+      *d_aggregate_out = reduce_op(*d_previous_aggregate, promote_op(per_partition_aggregate));
+    }
+
+    // If this is the last partition, we write the global aggregate to the user-provided iterator
+    if (last_partition)
+    {
+      *d_out = *d_aggregate_out;
+    }
+  }
+
+  /**
+   * This is a helper function that's invoked after a partition has been fully processed
+   */
+  template <typename GlobalOffsetT>
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void advance(GlobalOffsetT partition_size, bool next_partition_is_the_last)
+  {
+    promote_op.advance(partition_size);
+    ::cuda::std::swap(d_previous_aggregate, d_aggregate_out);
```

Comment:
Remark: the common pattern to swap anything is the two-step ADL dance:
```suggestion
    using ::cuda::std::swap;
    swap(d_previous_aggregate, d_aggregate_out);
```
It's not necessary here because the objects to be swapped are both raw pointers. You may still want to use the two-step in case we ever change the type of the pointers to anything that might bring in a custom swap (like a smart pointer).

Reply:
Thanks, I hadn't thought of that scenario! TIL.